src/pip/_internal/utils/pyc_compile.py

```diff
@@ -0,0 +1,170 @@
+            self.pool = futures.InterpreterPoolExecutor(workers)
+        else:
+            self.pool = futures.ProcessPoolExecutor(workers)
+        self.workers = workers
+
+    def __call__(self, paths: Iterable[Union[str, Path]]) -> Iterable[CompileResult]:
+        # New workers can be started at any time, so patch until fully done.
+        with _patch_main_module_hack():
+            yield from self.pool.map(_compile_single, paths)
+
+    def __exit__(self, *args: object) -> None:
+        # It's pointless to block on pool finalization, let it occur in background.
+        self.pool.shutdown(wait=False)
+
+
+def create_bytecode_compiler(
+    max_workers: WorkerSetting = "auto",
+    code_size_check: Optional[Callable[[int], bool]] = None,
+) -> BytecodeCompiler:
+    """Return a bytecode compiler appropriate for the workload and platform.
+
+    Parallelization will only be used if:
+      - There are 2 or more CPUs available
+      - The maximum # of workers permitted is at least 2
+      - There is "enough" code to be compiled to offset the worker startup overhead
+          (if it can be determined in advance via code_size_check)
+
+    A maximum worker count of "auto" will use the number of CPUs available to the
+    process or system, up to a hard-coded limit (to avoid resource exhaustion).
+
+    code_size_check is a callable that receives the code size threshold (in # of
+    bytes) for parallelization and returns whether it will be surpassed or not.
+    """
+    import logging
+
+    try:
+        # New in Python 3.13.
+        cpus: Optional[int] = os.process_cpu_count()  # type: ignore
+    except AttributeError:
+        # Poor man's fallback. We won't respect PYTHON_CPU_COUNT, but the envvar
+        # was only added in Python 3.13 anyway.
+        try:
+            cpus = len(os.sched_getaffinity(0))  # exists on unix (usually)
+        except AttributeError:
+            cpus = os.cpu_count()
+
+    logger = logging.getLogger(__name__)
+    logger.debug("Detected CPU count: %s", cpus)
+    logger.debug("Configured worker count: %s", max_workers)
+
+    # Case 1: Parallelization is disabled or pointless (there's only one CPU).
+    if max_workers == 1 or cpus == 1 or cpus is None:
+        logger.debug("Bytecode will be compiled serially")
+        return SerialCompiler()
+
+    # Case 2: There isn't enough code for parallelization to be worth it.
+    if code_size_check is not None and not code_size_check(CODE_SIZE_THRESHOLD):
+        logger.debug("Bytecode will be compiled serially (not enough .py code)")
+        return SerialCompiler()
+
+    # Case 3: Attempt to initialize a parallelized compiler.
+    # The concurrent executors will spin up new workers on a "on-demand basis",
+    # which helps to avoid wasting time on starting new workers that won't be
+    # used. (** This isn't true for the fork start method, but forking is
+    # fast enough that it doesn't really matter.)
```

Comment:
This could be smarter as the stdlib obviously doesn't consider the overhead of creating a new worker, but this on-demand behaviour is good enough (and manually scaling would be way too complicated).